server/src/test/java/io/crate/expression/predicate/DistinctFromTest.java

```diff
@@ -0,0 +1,216 @@
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
+
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.Version;
+import org.junit.Test;
+
+import io.crate.expression.scalar.ScalarTestCase;
+import io.crate.expression.symbol.Literal;
+import io.crate.expression.symbol.format.Style;
+import io.crate.lucene.GenericFunctionQuery;
+import io.crate.testing.DataTypeTesting;
+import io.crate.testing.QueryTester;
+import io.crate.types.ArrayType;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+
+/*
+A <distinct predicate> tests two values to see whether they are distinct and
+returns either ``TRUE`` or ``FALSE``.
+ */
+public class DistinctFromTest extends ScalarTestCase {
+
+    public static final List<Integer> ONE_TEN_ARRAY = List.of(1, 10);
+    public static final List<Integer> TWO_TWENTY_ARRAY = List.of(2, 20);
+
+    // * The two expressions must be comparable. If the comparands are rows, they
+    //  must be of the same degree and each corresponding pair of Fields must have
+    //  comparable <data type>s.
+    //
+    @Test
+    public void testEvaluateIncomparableDatatypes() {
+        assertThatThrownBy(() -> assertEvaluate("3 is distinct from true", isLiteral(false)))
+            .isExactlyInstanceOf(ClassCastException.class);
+    }
+
+    @Test
+    public void testNormalizeSymbolNullNull() {
+        // two ``NULL`` values are not distinct from one other
+        assertNormalize("null is distinct from null", isLiteral(Boolean.FALSE));
+    }
+
+    @Test
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    public void test_random_data_type() {
+        DataType<?> type = DataTypeTesting.randomTypeExcluding(Set.of(DataTypes.GEO_SHAPE));
+        Supplier<?> dataGenerator = getDataGenerator(type);
+
+        Object value1 = dataGenerator.get();
+        Object value2 = dataGenerator.get();
+        // Make sure the two values are distinct. Important for boolean or BitStringType.
+        while (((DataType) type).compare(value1, value2) == 0) {
+            value2 = dataGenerator.get();
+        }
+
+        Literal<?> literal1 = Literal.ofUnchecked(type, value1);
+        assertEvaluate("? IS NOT DISTINCT FROM ?", Boolean.TRUE, literal1, literal1);
+
+        Literal<?> literal2 = Literal.ofUnchecked(type, value2);
+        assertEvaluate("? IS DISTINCT FROM ?", Boolean.TRUE, literal1, literal2);
+    }
+
+    @Test
+    public void test_random_data_type_against_null() {
+        DataType<?> randomType = randomType();
+        var val = Literal.ofUnchecked(randomType, getDataGenerator(randomType).get()).toString(Style.QUALIFIED);
+        // Random value IS NOT distinct from null.
+        assertEvaluate("? IS NOT DISTINCT FROM null", Boolean.FALSE, Literal.of(val));
+        assertEvaluate("null IS NOT DISTINCT FROM ?", Boolean.FALSE, Literal.of(val));
```

Comment:
Sorry for nitpicking, `IS NOT DISTINCT` is basically `NOT(IS DISTINCT)` so the below two asserts are enough.

Reply:
Sure, I removed those. I just would not assume that NOT does always get properly evaluated from experience...